### PR TITLE
Multiple indexed commands 2

### LINKED ISSUE #110: Command that uses INDEX should allow more than one task to be modified at the same time
This is strictly optional though.

### LINKED ISSUE #122: Do something about completed tasks and events slowly building up
When using the application, I noticed that over time the tasks will start building up. This clutters up the app, and I think was the concern Damith expressed in the v0.2 evaluation. There's two approach 
- Remove completed tasks automatically 
- Add a clear command for the user to delete all completed tasks all at once 

I prefer the second approach - the first would mean that it would no longer be possible to use the app to measure eg. how much of a project is completed vs. how much more to do (ie as a burndown meter)

We could also have a configurable option that allows the first, but that would be an extension.

### PR DESCRIPTION
**Pull request by cricketer94**
Fixes #122. Can be extended to fix #110 as well.

### CODE REVIEW

File: src/main/java/seedu/todo/logic/commands/TagCommand.java [hunk 1]
```diff
@@ -120,6 +124,8 @@ public CommandResult execute() throws ValidationException {
         //Performs the actual execution with the data
         if (isAddTagsToTask()) {
             model.addTagsToTask(displayedIndex, tagsToAdd);
+            ImmutableTask task = model.getObservableList().get(displayedIndex - INDEX_OFFSET);
```
Comment (li-kai) [inline src/main/java/seedu/todo/logic/commands/TagCommand.java:218]: Use `model.getTask(displayedIndex)`

File: src/main/java/seedu/todo/model/TodoList.java [hunk 2]
```diff
@@ -99,50 +99,58 @@ public ImmutableTask add(String title, Consumer<MutableTask> update) throws Vali
...
         return task;
     }
 
     @Override
-    public ImmutableTask delete(int index) throws ValidationException {
-        Task task = tasks.remove(index);
+    public List<ImmutableTask> delete(List<Integer> indexes) throws ValidationException {
+        List<ImmutableTask> tasksRemoved = new ArrayList<>();
+        
+        for (Integer i : indexes) {
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/TodoList.java]: `for (Integer index : indexes)`, no need to rename the param 😄

File: src/main/java/seedu/todo/model/TodoList.java [hunk 3]
```diff
@@ -99,50 +99,58 @@ public ImmutableTask add(String title, Consumer<MutableTask> update) throws Vali
...
-        update.accept(validationTask);
-        validationTask.validate();
-
-        // changes are validated and accepted
-        update.accept(task);
-        saveTodoList();
-        return task;
+    
+    public ImmutableTask delete(int index) throws ValidationException {
+        List<Integer> indexes = new ArrayList<>();
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/TodoList.java]: Use the array instantiation shorthand here. And explain why `get(0)` was used.

File: src/main/java/seedu/todo/model/TodoList.java [hunk 4]
```diff
@@ -99,50 +99,58 @@ public ImmutableTask add(String title, Consumer<MutableTask> update) throws Vali
...
+        
         for (Integer x: indexes) {
             MutableTask task = tasks.get(x);
             ValidationTask validationTask = new ValidationTask(task);
             update.accept(validationTask);
             validationTask.validate();
         }
-        
+        List<ImmutableTask> tasksUpdated = new ArrayList<>();
         for (Integer i : indexes) {
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/TodoList.java]: Same problem, just rename the param in the for loop.

Explain why you're doing a for loop a second time (preventing partial updates)

File: src/main/java/seedu/todo/model/TodoModel.java [hunk 5]
```diff
@@ -151,26 +146,34 @@ public ImmutableTask delete(int index) throws ValidationException {
         uniqueTagCollection.notifyTaskDeleted(taskToDelete);
         return todoList.delete(taskIndex);
     }
+    
+    @Override
+    public List<ImmutableTask> deleteAll() throws ValidationException{
+        saveUndoState();
+        List<Integer> indexes = new ArrayList<>();
+        for (int i = 1 ; i <= getObservableList().size() ; i++) {
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/TodoModel.java]: [No spaces before semicolons](https://oss-generic.github.io/process/codingstandards/coding-standards-java.html#loops)

File: src/test/java/seedu/todo/logic/commands/ClearCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+public class ClearCommandTest extends CommandTest {
+    
+    private ImmutableTask task1, task2, task3;
+
+    @Before
+    public void setUp() throws Exception {
+        task1 = model.add("Hello", task -> {task.setCompleted(true); 
+                                    task.setEndTime(LocalDateTime.now().plusHours(2));
+                                    task.setPinned(true);
+                                    } );
```
Comment (li-kai) [inline src/test/java/seedu/todo/logic/commands/ClearCommandTest.java]: Mind your spaces `} )`. There's another one below, with wrong indentation too.

File: src/main/java/seedu/todo/model/TodoModel.java [hunk 7]
```diff
@@ -151,26 +146,34 @@ public ImmutableTask delete(int index) throws ValidationException {
...
+            indexes.add(getTaskIndex(i));
+            ImmutableTask taskToDelete = tasks.get(getTaskIndex(i));
+            uniqueTagCollection.notifyTaskDeleted(taskToDelete);
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/TodoModel.java]: What happens if a validation error occurs? Not all tasks are deleted now... 😆 
Only inform tags after all tasks are deleted.
Do it what you previously did (second for loop).

**Review by li-kai — CHANGES_REQUESTED** [review]
Just some stylistic stuff. Otherwise looks pretty straightforward.

File: src/main/java/seedu/todo/logic/commands/TagCommand.java (hunk 1 above)
Comment (cricketer94) [reply src/main/java/seedu/todo/logic/commands/TagCommand.java:218]: Whose branch is this on ah? Can't find leh

File: src/main/java/seedu/todo/model/TodoList.java (hunk 2 above)
Comment (cricketer94) [reply src/main/java/seedu/todo/model/TodoList.java]: The use of i is generally the convention used when iterating through integers though :x

File: src/main/java/seedu/todo/logic/commands/TagCommand.java (hunk 1 above)
Comment (li-kai) [reply src/main/java/seedu/todo/logic/commands/TagCommand.java:218]: Xien Dong's branch has it...you called this method somewhere

File: src/main/java/seedu/todo/model/TodoList.java (hunk 2 above)
Comment (li-kai) [reply src/main/java/seedu/todo/model/TodoList.java]: Yes it's semantic in a for(int i = 0)... kind of loop, but not this.

File: src/main/java/seedu/todo/logic/commands/TagCommand.java (hunk 1 above)
Comment (cricketer94) [reply src/main/java/seedu/todo/logic/commands/TagCommand.java:218]: Wrong model? Might only exist on the underlying layer. This model refers to the visible layer.

File: src/main/java/seedu/todo/logic/commands/ClearCommand.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public List<CommandSummary> getCommandSummary() {
+        return ImmutableList.of(new CommandSummary("Clear tasks in current view", getCommandName()));
+    }
+
+    @Override
+    public CommandResult execute() throws ValidationException {
+        model.deleteAll();
```
Comment (ZhangYiJiang) [inline src/main/java/seedu/todo/logic/commands/ClearCommand.java:31]: Uh... I thought the `clear` command is a solution to #122 - specifically this was _not_ meant to be a delete all command, but rather all _completed_ tasks

File: src/test/java/seedu/todo/logic/commands/ClearCommandTest.java (hunk 6 above)
Comment (ZhangYiJiang) [reply src/test/java/seedu/todo/logic/commands/ClearCommandTest.java]: This is not the correct way to indent lambdas. Treat the `{ }` like any other braces - 

``` java
model.add("Hello", task -> {
    /// ... 
});
```

File: src/main/java/seedu/todo/logic/commands/ClearCommand.java (hunk 8 above)
Comment (cricketer94) [reply src/main/java/seedu/todo/logic/commands/ClearCommand.java:31]: So do i prevent execution if the current view contains tasks which are not completed?

File: src/main/java/seedu/todo/logic/commands/TagCommand.java (hunk 1 above)
Comment (li-kai) [reply src/main/java/seedu/todo/logic/commands/TagCommand.java:218]: In that case, yeah, might be the wrong model 😅

**Review by li-kai — APPROVED** [review]

**Pull request merged**